### ISSUE TITLE
Previously opened project(s) not opens during a Moonshine restart

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#786]
So I noticed this problem during my tests on #756 . It took me some time to understand the correct steps in which scenario this problem triggers, and where it's happening. A test can be done using the master nightly build.

1. Close all the projects in current workspace (better not switch workspace as the steps involve switching workspace in later steps)
2. Restart Moonshine
3. The current workspace should populate with 0 projects (i.e. IDE-Default)
4. Open following 5 projects as an example:
- InstallerSharedCore
- MoonshineGUICore
- MoonshineSharedCore
- Native Royale
- VisualEditorConverterLib

5. At this point if one inspect the "moonshine-ide-project.sol" SharedObject file, s/he can see 5 projects' references correctly stored in the file
6. Create a new workspace - NewWorkspace
7. Re-open 3 projects from the above projects list:
- InstallerSharedCore
- MoonshineGUICore
- MoonshineSharedCore

8. Inspect the "moonshine-ide-project.sol" SharedObject file, s/he can see these 3 projects' references correctly stored in the file
9. A "Switch Workspace" to previous IDE-Default workspace properly re-opens all the 5 projects again, in the sidebar
10. At this point if inspects "moonshine-ide-project.sol" SharedObject file, it has reference of only 2 files - NativeRoyale and VisualEditorConverterLib - which were not present in NewWorkspace
11. Restarting Moonshine opens following 2 projects instead of 5 [Bug]:
- Native Royale
- VisualEditorConverterLib

So, in short, if a project is opened in A workspace, and that project is also exists in B workspace, changing to B workspace do not stores its reference anymore in the "moonshine-ide-project.sol" file.

**Comment by rat-moonshine** [comment]
As my investigation goes, it seems [hasItemForOpen](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/utils/SharedObjectUtil.as#L280) field set _true_ in case of a project that was already opened in a previous workspace before a switch. The field value therefore not saves the reference in the "moonshine-ide-project.sol" file anymore.

@piotrzarzycki21 what this check does and what will happen if this check is being removed?

**Comment by piotrzarzycki21** [comment]
I don't think it should be removed. I would rather go with implementing things in a way where saving projects for reopening taking into account Workspaces. When I was implemented that stuff a while ago workspaces didn't exists. 

Whole logic actually behave partially correctly if you think more about that. Cause initially 5 project opened has been stored, later 3 out of that 5 in different workspace. - Probably in the moment when you are in this new workspace you have actually 5 projects stored in the file - when you switch to old workspace 3 were removed, so 2 old one left... 

We need to adjust logic to have workspaces and inside those workspaces arrays of projects.

**Comment by rat-moonshine** [comment]
Yeah, that thought also hits me. But an initial look into the workspace plugin codes it tells me that all projects closed during a switch workspace, that expects removal of all the projects from the sidebar before re-opening the projects from the switching workspace again. So at the time of opening the new workspace's projects `hasItemForOpen` suppose to see it false.

Unless, a race-situation where the workspace plugin has dispatched projects' close event but not closed yet, and re-opening of projects hit. But this may unlikely also since the the reproducible steps are always right. 

This may need some more investigation where it's cooking, but I wanted to know what does this `hasItemForOpen` logic tests.

**Comment by rat-moonshine** [comment]
Following is some observation in nature, after some experience shared by @JoelProminic .

While testing on a VD environment of Windows 10, he experienced results during switching workspace, while,
IDE-Default - **Opened 10 projects**
Another Workspace - **Opened 0 projects**

Switch | Time Taken Till Populating the Sidebar
---------------- | ----------------
IDE-Default --> Another Workspace | 30 seconds
Another Workspace --> IDE-Default | 15 seconds

While the VD environment is generally slow, we wanted to know if the performance improvement still triggers during workspace changes or not; When I worries this slowness somehow affects by the projects' individual language-servers. 

To have my doubt test-out, I took the following tests using the branch build on regular Windows environment.

IDE-Default - **Opened 11 projects**
Another Workspace - **Opened 0 projects**
Java Available - **YES**

Attempt# | Switch | Time Taken Till Populating the Sidebar
---- | ---------------- | ----------------
Always | IDE-Default --> Another Workspace | 0 seconds
1-3 | Another Workspace --> IDE-Default | 5-10 seconds
4-6 | Another Workspace --> IDE-Default | >14 & <20 seconds
6+ | Another Workspace --> IDE-Default | >20 & <30 seconds

Java Available - **NO** (I require to ensure that either OpenJDK or "Program Files" Java is not auto-set in Moonshine)

Attempt# | Switch | Time Taken Till Populating the Sidebar
---- | ---------------- | ----------------
Always | IDE-Default --> Another Workspace | 0 seconds
>6 | Another Workspace --> IDE-Default | 4 seconds

The time taking to open large number of projects with language-server support increase gradually between workspace changes. This could possible because of AIR memory during a single session. I don't know where the memory holding but this could be a subject of investigation. 

In general, I don't see performance improvement creating itself any memory problem during the workspace changes, though.

**Comment by JoelProminic** [comment]
I did some tests with switching between workspaces with shared projects and restarting Moonshine, and I wasn't able to reproduce the original problem reported in this issue.
